### PR TITLE
Retry failed tasks

### PR DESCRIPTION
**Pull request by ruchim**
#3161 

- [x] Runtime attribute `maxRetries` implemented
- [x] Centaur tests
- [x] Changelog/Cromwell docs
- [x] More unit tests

### CODE REVIEW

File: backend/src/main/scala/cromwell/backend/BackendLifecycleActorFactory.scala [hunk 1]
```diff
@@ -115,9 +115,18 @@ trait BackendLifecycleActorFactory {
...
+
   /*
    * Returns credentials that can be used to authenticate to a docker registry server
    * in order to obtain a docker hash.
    */
   def dockerHashCredentials(initializationDataOption: Option[BackendInitializationData]): List[Any] = List.empty
 }
+
+object BackendLifecycleActorFactory {
+  val failedRetryCountKey = "FailedRetryCount"
```
Comment (kshakir) [inline backend/src/main/scala/cromwell/backend/BackendLifecycleActorFactory.scala]: `failedRetryCountKey` should be `FailedRetryCountKey`

File: backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala [hunk 2]
```diff
@@ -557,9 +571,9 @@ trait StandardAsyncExecutionActor extends AsyncBackendJobExecutionActor with Sta
     * Returns true if the status represents a success.
     *
     * @param runStatus The run status.
-    * @return True if the job is a success.
+    * @return True if the job is a done.
```
Comment (kshakir) [inline backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala]: `True if the job is a done.` should be `True if the job is done.`

File: backend/src/test/scala/cromwell/backend/validation/RuntimeAttributesValidationSpec.scala [hunk 3]
```diff
@@ -367,5 +367,15 @@ class RuntimeAttributesValidationSpec extends WordSpecLike with Matchers with Be
 
       ContinueOnReturnCodeValidation.configDefaultWdlValue(optinalBackendConfig).get shouldBe WomArray(WomArrayType(WomIntegerType), Array(WomInteger(0), WomInteger(1), WomInteger(2)))
     }
+
+    "return failure when tries to validate an invalid maxRetries entry" in {
+      val maxRetries = Some(WomInteger(-1))
```
Comment (kshakir) [inline backend/src/test/scala/cromwell/backend/validation/RuntimeAttributesValidationSpec.scala]: `Option(WomInteger(-1))`

File: server/src/test/scala/cromwell/SimpleWorkflowActorSpec.scala [hunk 4]
```diff
@@ -149,6 +149,7 @@ class SimpleWorkflowActorSpec extends CromwellTestKitWordSpec with BeforeAndAfte
       probe.expectTerminated(workflowActor, 2.seconds)
       supervisor.expectMsgPF(AwaitAlmostNothing, "parent should get a failed response") {
         case x: WorkflowFailedResponse =>
+          println(s"${x.reasons.head.getMessage}")
```
Comment (kshakir) [inline server/src/test/scala/cromwell/SimpleWorkflowActorSpec.scala]: Delete this `println`.

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PreviousRetryReasons.scala [hunk 5]
```diff
@@ -1,17 +1,19 @@
-package cromwell.backend.google.pipelines.common
+package cromwell.backend.impl.jes
```
Comment (kshakir) [inline supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PreviousRetryReasons.scala]: This looks like a bad rebase? The package doesn't match the file path.

File: backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala [hunk 6]
```diff
@@ -890,26 +924,31 @@ trait StandardAsyncExecutionActor extends AsyncBackendJobExecutionActor with Sta
...
+                retryElseFail(status, FailedNonRetryableExecutionHandle(StderrNonEmpty(jobDescriptor.key.tag, stderrSize, stderrAsOption), Option(returnCodeAsInt)))
               case Success(returnCodeAsInt) if isAbort(returnCodeAsInt) =>
                 Future.successful(AbortedExecutionHandle)
               case Success(returnCodeAsInt) if !continueOnReturnCode.continueFor(returnCodeAsInt) =>
-                Future.successful(FailedNonRetryableExecutionHandle(WrongReturnCode(jobDescriptor.key.tag, returnCodeAsInt, stderrAsOption), Option(returnCodeAsInt)))
+                retryElseFail(status, FailedNonRetryableExecutionHandle(WrongReturnCode(jobDescriptor.key.tag, returnCodeAsInt, stderrAsOption), Option(returnCodeAsInt)))
               case Success(returnCodeAsInt) =>
                 handleExecutionSuccess(status, oldHandle, returnCodeAsInt)
               case Failure(_) =>
+                //What does it mean that there is no return code-- should this be retried??
```
Comment (Horneth) [inline backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala]: It means for some reason the return code couldn't be parsed from String to Int for some reason, which would likely mean we messed up the Cromwell script. I don't thing this should be retried

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala [hunk 7]
```diff
@@ -525,22 +523,22 @@ class PipelinesApiAsyncBackendJobExecutionActor(override val standardParams: Sta
         val thisUnexpectedRetry = ur + 1
         if (thisUnexpectedRetry <= maxUnexpectedRetries) {
           // Increment unexpected retry count and preemption count stays the same
-          writeFuturePreemptedAndUnexpectedRetryCounts(p, thisUnexpectedRetry).map { _ =>
```
Comment (Horneth) [inline supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala]: OOC: why the change about not waiting anymore for the result of `writeFuturePreemptedAndUnexpectedRetryCounts` ?

File: CHANGELOG.md [hunk 8]
```diff
@@ -24,6 +24,9 @@ Please update your configuration accordingly.
 If you don't update the `actor-factory` value, you'll get a deprecation warning in the logs, and Cromwell will default back to
 **PAPI V1**
 
+### Task Retries
+Cromwell now supports retrying failed tasks upto a specified count by declaring a value for the [maxRetries](RuntimeAttributes.md#maxRetries) key through the WDL runtime attributes.
```
Comment (Horneth) [inline CHANGELOG.md]: `upto` -> `up to` ?

File: backend/src/main/scala/cromwell/backend/BackendLifecycleActorFactory.scala [hunk 9]
```diff
@@ -115,9 +115,18 @@ trait BackendLifecycleActorFactory {
     */
   def requestedKeyValueStoreKeys: Seq[String] = Seq.empty
 
+  /**
+    * A set of KV store keys that are requested on behalf of all backends that the engine lookup before running each job.
```
Comment (mcovarr) [inline backend/src/main/scala/cromwell/backend/BackendLifecycleActorFactory.scala]: the sentence is off

File: backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala [hunk 10]
```diff
@@ -446,6 +446,20 @@ trait StandardAsyncExecutionActor extends AsyncBackendJobExecutionActor with Sta
   lazy val continueOnReturnCode: ContinueOnReturnCode = RuntimeAttributesValidation.extract(
     ContinueOnReturnCodeValidation.instance, validatedRuntimeAttributes)
 
+  /**
+    * Returns the max number of times that a failed job should be retried, obtained by converting `maxRetries` to an Int.
+    *
+    * @return the behavior for max retries.
```
Comment (mcovarr) [inline backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala]: I'd delete this line, it's just kind of confusing

File: backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala (hunk 2 above)
Comment (mcovarr) [reply backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala]: or `True if the job is a done deal.`

File: backend/src/main/scala/cromwell/backend/validation/MaxRetriesValidation.scala [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+  * reference.conf file, coerced into a WomValue.
+  */
+object MaxRetriesValidation {
+  lazy val instance: RuntimeAttributesValidation[Int] = new MaxRetriesValidation(MaxRetriesKey)
+  lazy val optional: OptionalRuntimeAttributesValidation[Int] = instance.optional
+
+  def default(runtimeConfig: Option[Config]): RuntimeAttributesValidation[Int] = instance.withDefault(
+    configDefaultWomValue(runtimeConfig) getOrElse WomInteger(0))
+  def configDefaultWomValue(config: Option[Config]): Option[WomValue] = instance.configDefaultWomValue(config)
+  //maybe this needs a cieling declared? It could be detrimental....
```
Comment (mcovarr) [inline backend/src/main/scala/cromwell/backend/validation/MaxRetriesValidation.scala]: i before e except after c

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/job/EngineJobExecutionActor.scala [hunk 12]
```diff
@@ -674,7 +675,7 @@ class EngineJobExecutionActor(replyTo: ActorRef,
       case SucceededResponseData(JobSucceededResponse(jobKey, returnCode, jobOutputs, _, _, _), hashes) =>
         publishHashesToMetadata(hashes)
         saveSuccessfulJobResults(jobKey, returnCode, jobOutputs)
-      case FailedResponseData(JobFailedNonRetryableResponse(jobKey, throwable, returnCode), hashes) =>
+      case FailedResponseData(JobFailedNonRetryableResponse(jobKey, throwable, returnCode), hashes)  =>
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/job/EngineJobExecutionActor.scala]: 🔙

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/job/preparation/JobPreparationActor.scala [hunk 13]
```diff
@@ -60,7 +60,7 @@ class JobPreparationActor(workflowDescriptor: EngineWorkflowDescriptor,
 
   when(Idle) {
     case Event(Start(valueStore), JobPreparationActorNoData) =>
-      evaluateInputsAndAttributes(valueStore) match {
+       evaluateInputsAndAttributes(valueStore) match {
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/job/preparation/JobPreparationActor.scala]: 🔙

File: backend/src/main/scala/cromwell/backend/validation/MaxRetriesValidation.scala (hunk 11 above)
Comment (ruchim) [reply backend/src/main/scala/cromwell/backend/validation/MaxRetriesValidation.scala]: Since nobody else mentioned a ceiling is required, I'll go ahead and delete the comment!

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala (hunk 7 above)
Comment (ruchim) [reply supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala]: Futures are back for the future.
